### PR TITLE
Fixed ordering in adding pending picks to RR

### LINKED ISSUE #15892: round robin load balancing on channels sharing sub-channels broken
### What version of gRPC and what language are you using?
```
$ python --version
Python 2.7.13

$ pip freeze  (selected lines)
grpcio==1.12.0
protobuf==3.5.2.post1

```
### What operating system (Linux, Windows, …) and version?
```
$ uname -a
Linux instance-3 4.9.0-6-amd64 #1 SMP Debian 4.9.88-1+deb9u1 (2018-05-07) x86_64 GNU/Linux
 
```
### What runtime / compiler are you using (e.g. python version or version of gcc)
```
$ python --version
Python 2.7.13

``` 
### What did you do?

I enable round robin load balancing on my python client.  I setup my channel as follows:

```
 opts = [("grpc.lb_policy_name", "round_robin",)]
 channel = grpc.insecure_channel('localhost:50051' , opts)

```

I updated the hello world example to repoduce this issue: 
https://github.com/tomhostyn/grpc/blob/round_robin_issue/examples/python/helloworld/greeter_client.py

I tried to reproduce the same issue in the cpp client, but that works as expected
https://github.com/tomhostyn/grpc/blob/round_robin_issue/examples/cpp/helloworld/greeter_client.cc

 
### What did you expect to see?

all messages sent 
 
### What did you see instead?

After sending 1 message succesfully, my client blocks.
when the modified example is executed, the output is as expected until :
```
========  with round robin =======
Connection OK
Greeter client received: Hello you
Connection OK
```

execution with verbose output 
```
$ env | grep 
GRPC
GRPC_TRACE=all
GRPC_VERBOSITY=DEBUG
$ python greeter_client.py &> log
^C^C^C

```
log output at https://pastebin.com/rxyp6huv

Closing the channel fixes the problem.  see discussion:
[](https://groups.google.com/forum/?utm_medium=email&utm_source=footer#!msg/grpc-io/cN_E2ASt5LA/uPBm6P7RAAAJ)

### Anything else we should know about your project / environment?
No

### PR DESCRIPTION
**Pull request by dgquintas**
Otherwise we may try to drain the pending picks _before_ the pending pick has been added. 

If READY subchannels are being recycled (for instance, if channels against the same target/addresses are being created and destroyed in a loop), the RR LB policy of a new channel will pick these subchannels in READY state. However, prior to this fix, `DrainPendingPicksLocked` was called _before_ the pending pick is registered: only connectivity state _transitions_ would cause the pending picks to be reprocessed. Which is why things seemed to work, as in most cases, an LB policy (and its subchannels) would start in a state other than READY and then transition into it.

Consider (with extra logging and whitespace added for clarity):

```
I0706 13:52:28.572841742  206522 round_robin.cc:358]         [RR 0x7fdd5c004790] Trying to pick (shutdown: 0)

// The following block encompases DoPick
I0706 13:52:28.572850644  206522 round_robin.cc:362]         [RR 0x7fdd5c004790] XXX before DoPick
I0706 13:52:28.572868100  206522 round_robin.cc:555]         [RR 0x7fdd5c004790] getting next ready subchannel (out of 2), last_ready_index=18446744073709551615
I0706 13:52:28.572882302  206522 round_robin.cc:564]         [RR 0x7fdd5c004790] checking subchannel 0x7fdd5c005f90, subchannel_list 0x7fdd5c006d50, index 0: state=IDLE
I0706 13:52:28.572892832  206522 round_robin.cc:564]         [RR 0x7fdd5c004790] checking subchannel 0x7fdd5c006930, subchannel_list 0x7fdd5c006d50, index 1: state=IDLE
I0706 13:52:28.572898391  206522 round_robin.cc:582]         [RR 0x7fdd5c006d50] no subchannels in ready state
I0706 13:52:28.572923655  206522 round_robin.cc:364]         [RR 0x7fdd5c004790] XXX after DoPick


// The following encompases what happens after DoPick() has returned false,
// indiating an async pick.

I0706 13:52:28.572929421  206522 round_robin.cc:366]         [RR 0x7fdd5c004790] YYY before adding pick to pending_picks_=(nil)
I0706 13:52:28.572938257  206522 connectivity_state.cc:92]   CONWATCH: 0x7fdd5c0060b0 subchannel: get READY
I0706 13:52:28.573123261  206522 round_robin.cc:508]         [RR 0x7fdd5c004790] connectivity changed for subchannel 0x7fdd5c005f90, subchannel_list 0x7fdd5c006d50 (index 0 of 2): prev_state=IDLE new_state=READ
Y
I0706 13:52:28.573133577  206522 connectivity_state.cc:92]   CONWATCH: 0x7fdd5c006a50 subchannel: get READY
I0706 13:52:28.573141425  206522 round_robin.cc:508]         [RR 0x7fdd5c004790] connectivity changed for subchannel 0x7fdd5c006930, subchannel_list 0x7fdd5c006d50 (index 1 of 2): prev_state=IDLE new_state=READ
Y

// At this point, we try to process pending picks, but note no pending pick has
// been added yet.

I0706 13:52:28.573148090  206522 round_robin.cc:344]         [RR 0x7fdd5c004790] XXX DrainPendingPicksLocked pending_picks_=(nil) 
I0706 13:52:28.573182049  206522 connectivity_state.cc:164]  SET: 0x7fdd5c0047e8 round_robin: IDLE --> READY [rr_ready] error=(nil) "No Error"
I0706 13:52:28.573187842  206522 connectivity_state.cc:190]  NOTIFY: 0x7fdd5c0047e8 round_robin: 0x7fdd5c002ce8
I0706 13:52:28.573207063  206522 subchannel_list.h:292]      [round_robin 0x7fdd5c004790] subchannel list 0x7fdd5c006d50 index 0 of 2 (subchannel 0x7fdd5c005f90): starting watch: requesting connectivity change 
notification (from READY)
I0706 13:52:28.573250859  206522 subchannel_list.h:292]      [round_robin 0x7fdd5c004790] subchannel list 0x7fdd5c006d50 index 1 of 2 (subchannel 0x7fdd5c006930): starting watch: requesting connectivity change notification (from READY)

// It is at this point that the pending pick is registered, by the attempt to
// drain the pending picks has already happened and the policy is READY, which
// means the (pending) pick will stall until something else happens to the
// subchannels/policy (eg. disconnection).
I0706 13:52:28.573297056  206522 round_robin.cc:373]         [RR 0x7fdd5c004790] YYY after adding pick to pending_picks_=0x7fdd680266f8
```

Fixes #15892

### CODE REVIEW

**Comment by dgquintas** [comment]
FYI, I'll add a new test for this shortly.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  +0.0%      +8 [None]    +256  +0.0%

  +0.0%      +8 TOTAL     +256  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  +0.0%      +8 [None]    +248  +0.0%

  +0.0%      +8 TOTAL     +248  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by dgquintas** [comment]
Test added. I've verified that it fails without this PR's fix.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  +0.0%      +8 [None]    +248  +0.0%

  +0.0%      +8 TOTAL     +248  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  +0.0%      +8 [None]    +248  +0.0%

  +0.0%      +8 TOTAL     +248  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: test/cpp/end2end/client_lb_end2end_test.cc [hunk 1]
```diff
@@ -539,6 +539,33 @@ TEST_F(ClientLbEnd2endTest, RoundRobin) {
...
+TEST_F(ClientLbEnd2endTest, RoundRobinProcessPending) {
+  StartServers(1);  // Single server
+  auto channel = BuildChannel("round_robin");
+  auto stub = BuildStub(channel);
+  SetNextResolution({servers_[0]->port_});
+  WaitForServer(stub, 0, DEBUG_LOCATION);
+  constexpr int kIterations = 4;
+  constexpr int kNumThreads = 4;
+  std::vector<std::thread> threads;
+  // Create and destroy several channels concurrently, executing an RPC each
```
Comment (markdroth) [inline test/cpp/end2end/client_lb_end2end_test.cc]: We can probably trigger this without multiple threads once #15841 is merged.  Maybe add a TODO assigned to @AspirinSJL?

**Comment by AspirinSJL** [comment]
If the subchannel is READY and used by a client, will a new client with the same target pick that subchannel? I don't see any reason for no. So I am wondering why we should recycle the subchannel to trigger this issue.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  +0.0%      +8 [None]    +256  +0.0%

  +0.0%      +8 TOTAL     +256  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by dgquintas** [comment]
Because by recycling the subchannel, the RR policy is "born READY", so there's no transition triggering a new drainage of pending picks. Basically we need to "warm up" the subchannels somewhere else prior to a new RR policy picking them up, resulting in a RR initial state other than the usual IDLE/CONNECTING

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by AspirinSJL** [comment]
Yeah, I can understand that the subchannel should be born READY to trigger this issue. I don't understand why we should recycle the subchannel though. Isn't the subchannel still READY even without recycling?

**Comment by dgquintas** [comment]
Maybe "recycle" isn't the right word. Yes, we use the same subchannel, which stays READY the whole time. Reuse is probably a better description.

File: test/cpp/end2end/client_lb_end2end_test.cc (hunk 1 above)
Comment (AspirinSJL) [reply test/cpp/end2end/client_lb_end2end_test.cc]: If we don't need to recycle the subchannel to trigger the issue, wouldn't it be sufficient to start one client, wait until its subchannel is READY, then start a second client with the same target? The second client will have a born READY subchannel in this case, right?

Comment (dgquintas) [reply test/cpp/end2end/client_lb_end2end_test.cc]: Probably. However, the issue is time-sensitive, so using threads and a loop makes it much more likely to happen. For testing, the more deterministic the failure, the better.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  +0.0%      +8 [None]    +248  +0.0%

  +0.0%      +8 TOTAL     +248  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by dgquintas** [comment]
Thank you @AspirinSJL for not settling for an unsatisfactory explanation and keep digging. This has resulted in a much simpler test.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  +0.0%      +8 [None]    +248  +0.0%

  +0.0%      +8 TOTAL     +248  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

File: test/cpp/end2end/client_lb_end2end_test.cc [hunk 2]
```diff
@@ -539,6 +539,25 @@ TEST_F(ClientLbEnd2endTest, RoundRobin) {
...
 }
 
+TEST_F(ClientLbEnd2endTest, RoundRobinProcessPending) {
+  StartServers(1);  // Single server
+  auto channel = BuildChannel("round_robin");
+  auto stub = BuildStub(channel);
+  SetNextResolution({servers_[0]->port_});
+  WaitForServer(stub, 0, DEBUG_LOCATION);
+  // Create a new channel and its corresponding RR LB policy, which will pick
+  // the subchannels in READY state from a the previous RPC against the same
```
Comment (AspirinSJL) [inline test/cpp/end2end/client_lb_end2end_test.cc]: Nit: "a the" -> "the"

File: test/cpp/end2end/client_lb_end2end_test.cc [hunk 3]
```diff
@@ -539,6 +539,25 @@ TEST_F(ClientLbEnd2endTest, RoundRobin) {
...
+  // target (even if it happened over a different channel, because subchannels
+  // are globally reused). Progress should happen without any transition from
+  // this READY state.
+  {
```
Comment (AspirinSJL) [inline test/cpp/end2end/client_lb_end2end_test.cc]: Since the second channel has a different name now, this bracket can probably be removed.

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE              FILE SIZE
 ++++++++++++++ GROWIN ++++++++++++++
  +0.0%      +8 [None]    +248  +0.0%

  +0.0%      +8 TOTAL     +248  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Pull request merged**